lib/jxl/modular/transform/squeeze.cc

```diff
@@ -43,36 +185,105 @@ Status InvHSqueeze(Image &input, uint32_t c, uint32_t rc, ThreadPool *pool) {
+  intptr_t onerow_inr = chin_residual.plane.PixelsPerRow();
+  intptr_t onerow_out = chout.plane.PixelsPerRow();
+  constexpr size_t kRowsPerThread = 8;
   JXL_RETURN_IF_ERROR(RunOnPool(
-      pool, 0, chin.h, ThreadPool::NoInit,
+      pool, 0, chin.h / kRowsPerThread, ThreadPool::NoInit,
+      [&](const uint32_t task, size_t /* thread */) {
+        const size_t y0 = task * kRowsPerThread;
+        const pixel_type *JXL_RESTRICT p_residual = chin_residual.Row(y0);
+        const pixel_type *JXL_RESTRICT p_avg = chin.Row(y0);
+        pixel_type *JXL_RESTRICT p_out = chout.Row(y0);
+
+        HWY_ALIGN pixel_type b_p_avg[9 * kRowsPerThread];
+        HWY_ALIGN pixel_type b_p_residual[8 * kRowsPerThread];
+        HWY_ALIGN pixel_type b_p_out_even[kRowsPerThread];
+        HWY_ALIGN pixel_type b_p_out_odd[kRowsPerThread];
+        size_t x = 0;
+        if (chin_residual.w > 16) {
+          for (; x < chin_residual.w - 9; x += 8) {
+            Transpose8x8Block(p_residual + x, b_p_residual, onerow_inr);
+            Transpose8x8Block(p_avg + x, b_p_avg, onerow_in);
+            for (size_t y = 0; y < kRowsPerThread; y++) {
+              b_p_avg[8 * 8 + y] = p_avg[x + 8 + onerow_in * y];
+            }
+            for (size_t i = 0; i < 8; i++) {
+              fast_unsqueeze<kRowsPerThread>(
+                  b_p_residual + 8 * i, b_p_avg + 8 * i, b_p_avg + 8 * (i + 1),
+                  (x + i ? b_p_out_odd : b_p_avg + 8 * i), b_p_out_even,
+                  b_p_out_odd);
+              for (size_t y = 0; y < kRowsPerThread; y++) {
+                p_out[((x + i) << 1) + onerow_out * y] = b_p_out_even[y];
+                p_out[((x + i) << 1) + 1 + onerow_out * y] = b_p_out_odd[y];
+              }
+            }
+          }
+        }
+        size_t x0 = x;
+        for (size_t y = 0; y < kRowsPerThread; y++) {
+          for (x = x0; x < chin_residual.w; x++) {
+            pixel_type diff_minus_tendency = p_residual[x + onerow_inr * y];
+            pixel_type avg = p_avg[x + onerow_in * y];
+            pixel_type next_avg =
+                (x + 1 < chin.w ? p_avg[x + 1 + onerow_in * y] : avg);
+            pixel_type left = (x ? p_out[(x << 1) - 1 + onerow_out * y] : avg);
+            pixel_type tendency = SmoothTendency(left, avg, next_avg);
+            pixel_type diff = diff_minus_tendency + tendency;
+            pixel_type A =
+                ((avg * 2) + diff + (diff > 0 ? -(diff & 1) : (diff & 1))) >> 1;
+            p_out[(x << 1) + onerow_out * y] = A;
+            pixel_type B = A - diff;
+            p_out[(x << 1) + 1 + onerow_out * y] = B;
+          }
+          if (chout.w & 1)
+            p_out[chout.w - 1 + onerow_out * y] =
+                p_avg[chin.w - 1 + onerow_in * y];
+        }
+      },
+      "InvHorizontalSqueeze"));
+  size_t firstrow = chin.h / kRowsPerThread * kRowsPerThread;
+#else
+  size_t firstrow = 0;
+#endif
+
+  JXL_RETURN_IF_ERROR(RunOnPool(
```

Comment:
I would either extract the one row scalar code to a function or get rid of this second case (it doesn't make that much of a difference, and you can fall back to scalar code if you don't have enough rows in the SIMD code).

Or both ;)

Reply:
Extracted it to a one row scalar unsqueeze function. It does have to be called in two spots because there's both the `height % 8` rows at the bottom and the `width % something` columns on the right that have to be done without simd.

Possibly at some point in the future if we have row padding in modular planes, the columns on the right will no longer require special handling.

Reply:
A solution to that would be having num_simd = 0 for last rows, (width/8*8) otherwise :)

Reply:
Did something like that. The second RunOnPool() is now gone.